core/discovery/broker/subscriber.go

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) 2019 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package broker
+
+import (
+	"sync"
+	"time"
+
+	"github.com/mysteriumnetwork/node/communication"
+	"github.com/mysteriumnetwork/node/communication/nats"
+	"github.com/mysteriumnetwork/node/core/discovery"
+	"github.com/mysteriumnetwork/node/market"
+)
+
+// proposalSubscriber responsible for handling proposal events through Broker (Mysterium Communication)
+type proposalSubscriber struct {
+	proposalStorage  *discovery.ProposalStorage
+	proposalIdleness time.Duration
+	receiver         communication.Receiver
+
+	watchdogStep time.Duration
+	watchdogStop chan bool
+	watchdogLock sync.Mutex
+	watchdogSeen map[market.ProposalID]time.Time
+}
+
+// NewProposalSubscriber returns new proposalSubscriber instance.
+func NewProposalSubscriber(
+	proposalsStorage *discovery.ProposalStorage,
+	proposalIdleness time.Duration,
+	connection nats.Connection,
+) *proposalSubscriber {
+	return &proposalSubscriber{
+		proposalStorage:  proposalsStorage,
+		proposalIdleness: proposalIdleness,
+		receiver:         nats.NewReceiver(connection, communication.NewCodecJSON(), "*"),
+
+		watchdogStep: proposalIdleness / 10,
```

Comment:
I would rather init my own interval here with some duration like 10 * time.Second etc. It is not clear why watchdogStep should depend on proposalIdleness. If it should maybe add comment for why 10 is chosen.

Reply:
Made injectable